projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/lineage_utils.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+import logging
+import re
+from typing import Optional
+from typing import Tuple
+
+from impala._thrift_gen.RuntimeProfile.ttypes import TRuntimeProfileFormat
+from impala.hiveserver2 import HiveServer2Cursor
+from vdk.api.lineage.model.sql.model import LineageData
+from vdk.api.lineage.model.sql.model import LineageTable
+
+_log = logging.getLogger(__name__)
+
+
+def get_lineage_data(cursor: HiveServer2Cursor) -> Optional[LineageData]:
+    """
+    Calculates lineage information from the query plan for the last executed query from the provided cursor. Lineage
```

Comment:
https://docs.cloudera.com/runtime/7.2.10/impala-reference/topics/impala-profile.html docs state:

> It shows physical details such as the number of bytes read, maximum memory usage, and so on for each node.

How is TGetRuntimeProfileReq->TGetRuntimeProfileResp evaluated after each query, in terms of load? What nodes do handle that request? 
What's the overhead, and is there a correlation with executor / coordinator nodes quantity (is handling a profile request heavier when having more cluster nodes)?

Reply:
The query engine builds a profile for every query that it executes regardless of if it is requested or not. TGetRuntimeProfileReq just fetches the already calculated profile so in terms of load it is mainly I/O. I believe this is handled by the coordinator that is responsible for executing the query. So the number of executors nodes in the impala cluster should not affect in any way this operation. The number of coordinator nodes in the cluster should not affect the overhead of a **single** get profile operation.

Reply:
The expectations regarding load and performance I imagine being documented in the plugin README.md, since it evaluates Impala cluster requests practically upon each query

Reply:
Ok, will do. But they cannot be very detailed as the performance aspect depends a lot on the actual load on the cluster we could not know.

Reply:
I looked at the README.md,
So we don't have any performance measurements made in advance? And, at the same time, I don't understand what request is made to Impala, and what do we ask to be performed exactly? What is the scope/workflow of the cluster components to be involved?
for example is `get_profile` handled by an executor that makes an RPC request to the coordinator, that keeps a profile forest, for how long, is it serialized, is it overwritten by keep alive queries performed ?

Now with the README.md, I don't have any more information at hand, just a brief statement that Impala is requested that I knew before as it's too obvious

Reply:
The README is too generic for me, and I don't have any impression nor understanding about what's going on under the hood; however we can ignore that, and merge so we have this capability available, right?

Reply:
I have linked Impala docs that answer some of the questions that you have and clarified the README.md to share a bit more details.
I have tested this functionality only against local containerised Impala and there was no measurable performance impact. However this is not representative as it is not run against big loaded environment. Once we have a release of the plugin will measure against more representable environment and share if there are any meaningful findings. 

Is get_profile handled by an executor that makes an RPC request to the coordinator, that keeps a profile forest, for how long, is it serialized?
The RPC originates from  the data job not Impala executor. How many profiles are kept is a cluster configuration, it is explained in the linked documents. They are serialized, again explained in the linked documents.

Is it overwritten by keep alive queries performed? No as the RPC is executed right after actual query is executed, keepalive queries are executed before actual queries are executed. This is elaborated in the README.md
The profile

Reply:
Let me know if this is enough?

Reply:
I guess